### PR TITLE
Add (symmetric) double-sided Crystal Ball shape to RooFit

### PR DESCRIPTION
**Pull request by goi42**
`RooCBShape` is commonly used in HEP analyses. This adds its slightly less common extrapolations `RooDSCBShape` and `RooSDSCBShape`.

### CODE REVIEW

**Comment by lmoneta** [comment]
@phsftbot build!

**Comment by lmoneta** [comment]
@phsft-bot build!

File: roofit/roofit/src/RooDSCBShape.cxx [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+ * @(#)root/roofit:$Id$
+ * Authors:                                                                  *
+ *    T. Skwarnicki modify RooCBShape to Asymmetrical Double-Sided CB                     *
+ *    Michael Wilkinson add to RooFit source                                 *
+ *****************************************************************************/
+
+/** \class RooCBShape
+    \ingroup Roofit
+
+PDF implementing the Asymmetrical Double-Sided Crystall Ball line shape.
```
Comment (hageboeck) [inline roofit/roofit/src/RooDSCBShape.cxx:13]: Could we add a latex equation here? See e.g.
https://github.com/root-project/root/blob/master/roofit/roofit/src/RooJohnson.cxx#L25

If you also happen to have a plot, I would add this once this is merged.

File: roofit/roofit/src/RooDSCBShape.cxx [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+#include "RooFit.h"
+
+#include "RooAbsReal.h"
+#include "RooRealVar.h"
+#include "RooMath.h"
+
+#include "TMath.h"
+
+#include "Riostream.h"
+#include "Riostream.h"
```
Comment (hageboeck) [inline roofit/roofit/src/RooDSCBShape.cxx:26]: We include `<iostream>` only if necessary in newer code.
```suggestion
#include <stdexcept>
```

File: roofit/roofit/src/RooDSCBShape.cxx [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+
+  return 0;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+Double_t RooDSCBShape::analyticalIntegral(Int_t code, const char* rangeName) const
+{
+  static const double sqrtPiOver2 = 1.2533141373;
+  static const double sqrt2 = 1.4142135624;
```
Comment (hageboeck) [inline roofit/roofit/src/RooDSCBShape.cxx:115]:
```suggestion
  constexpr double sqrtPiOver2 = 1.2533141373;
  constexpr double sqrt2 = 1.4142135624;
```

File: roofit/roofit/src/RooDSCBShape.cxx [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+    double term2 = sig*sqrtPiOver2*(   ApproxErf(absAlphaR/sqrt2)
+                                     - ApproxErf(tmin/sqrt2) );
+
+
+    result += term1 + term2;
+  }
+  else {
+    std::cout << " This should never happen! Error in RooDSCBShape logic " << std::endl;
+    assert(0);
```
Comment (hageboeck) [inline roofit/roofit/src/RooDSCBShape.cxx:229]: Asserts don't work in released code.
```suggestion
    throw std::logic_error("This should never happen! Error in RooDSCBShape logic. Please file a bug report.");
```

File: roofit/roofit/src/RooSDSCBShape.cxx [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+ * @(#)root/roofit:$Id$
+ * Authors:                                                                  *
+ *    T. Skwarnicki modify RooCBShape to Symmetrical Double-Sided CB         *
+ *    Michael Wilkinson add to RooFit source                                 *
+ *****************************************************************************/
+
+/** \class RooCBShape
+    \ingroup Roofit
+
+PDF implementing the Symmetrical Double-Sided Crystall Ball line shape.
```
Comment (hageboeck) [inline roofit/roofit/src/RooSDSCBShape.cxx:13]: Latex formula if possible.

File: roofit/roofit/src/RooSDSCBShape.cxx [hunk 6]
```diff
@@ -0,0 +1,246 @@
...
+#include "RooFit.h"
+
+#include "RooAbsReal.h"
+#include "RooRealVar.h"
+#include "RooMath.h"
+
+#include "TMath.h"
+
+#include "Riostream.h"
+#include "Riostream.h"
```
Comment (hageboeck) [inline roofit/roofit/src/RooSDSCBShape.cxx:26]:
```suggestion
#include <stdexcept>
```

File: roofit/roofit/src/RooSDSCBShape.cxx [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+    return 1.0;
+  if( arg < -erflim )
+    return -1.0;
+
+  return RooMath::erf(arg);
+}
+
+
+static const char rcsid[] =
+"$Id: RooSDSCBShape.C 1246 2013-04-02 02:51:06Z tskwarni $";
```
Comment (hageboeck) [inline roofit/roofit/src/RooSDSCBShape.cxx:48]:
```suggestion
```

File: roofit/roofit/src/RooSDSCBShape.cxx [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+
+  return 0;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+Double_t RooSDSCBShape::analyticalIntegral(Int_t code, const char* rangeName) const
+{
+  static const double sqrtPiOver2 = 1.2533141373;
+  static const double sqrt2 = 1.4142135624;
```
Comment (hageboeck) [inline roofit/roofit/src/RooSDSCBShape.cxx:107]:
```suggestion
  constexpr double sqrtPiOver2 = 1.2533141373;
  constexpr double sqrt2 = 1.4142135624;
```

File: roofit/roofit/src/RooSDSCBShape.cxx [hunk 9]
```diff
@@ -0,0 +1,246 @@
...
+
+    double term2 = sig*sqrtPiOver2*(   ApproxErf(-tmin/sqrt2)
+                                     - ApproxErf(-absAlpha/sqrt2) );
+
+
+    result += term1 + term2;
+  }
+  else {
+    std::cout << " This should never happen! Error in RooSDCCBShape logic " << std::endl;
+    assert(0);
```
Comment (hageboeck) [inline roofit/roofit/src/RooSDSCBShape.cxx:219]:
```suggestion
    throw std::logic_error("This should never happen! Error in RooSDCCBShape logic. Please file a bug report.");
```

File: roofit/roofit/src/RooSDSCBShape.cxx [hunk 10]
```diff
@@ -0,0 +1,246 @@
...
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+Double_t RooSDSCBShape::analyticalIntegral(Int_t code, const char* rangeName) const
+{
+  static const double sqrtPiOver2 = 1.2533141373;
+  static const double sqrt2 = 1.4142135624;
+
+  assert(code==1);
```
Comment (hageboeck) [inline roofit/roofit/src/RooSDSCBShape.cxx:109]: Suppress the unused warning in release code:
```suggestion
  assert(code==1);
  (void)code;
```

File: roofit/roofit/src/RooDSCBShape.cxx (hunk 1 above)
Comment (goi42) [reply roofit/roofit/src/RooDSCBShape.cxx:13]: done. I can create a plot after this is merged.

File: roofit/roofit/src/RooDSCBShape.cxx (hunk 2 above)
Comment (goi42) [reply roofit/roofit/src/RooDSCBShape.cxx]: done.

File: roofit/roofit/src/RooDSCBShape.cxx (hunk 3 above)
Comment (goi42) [reply roofit/roofit/src/RooDSCBShape.cxx]: done.

File: roofit/roofit/src/RooDSCBShape.cxx (hunk 4 above)
Comment (goi42) [reply roofit/roofit/src/RooDSCBShape.cxx]: done.

File: roofit/roofit/src/RooSDSCBShape.cxx (hunk 5 above)
Comment (goi42) [reply roofit/roofit/src/RooSDSCBShape.cxx:13]: done. I can create a plot after this is merged.

File: roofit/roofit/src/RooSDSCBShape.cxx (hunk 6 above)
Comment (goi42) [reply roofit/roofit/src/RooSDSCBShape.cxx]: done.

File: roofit/roofit/src/RooSDSCBShape.cxx (hunk 7 above)
Comment (goi42) [reply roofit/roofit/src/RooSDSCBShape.cxx]: done.

File: roofit/roofit/src/RooSDSCBShape.cxx (hunk 8 above)
Comment (goi42) [reply roofit/roofit/src/RooSDSCBShape.cxx]: done. I found also an additional `static const double` that I also replaced with `constexpr double`.

File: roofit/roofit/src/RooSDSCBShape.cxx (hunk 9 above)
Comment (goi42) [reply roofit/roofit/src/RooSDSCBShape.cxx]: done.

File: roofit/roofit/src/RooSDSCBShape.cxx (hunk 10 above)
Comment (goi42) [reply roofit/roofit/src/RooSDSCBShape.cxx]: done. Also changed to `R__ASSERT`.

File: roofit/roofit/src/RooDSCBShape.cxx (hunk 2 above)
Comment (goi42) [reply roofit/roofit/src/RooDSCBShape.cxx]: @hageboeck This inclusion seems to result in a build failure, claiming stdexcept cannot be found. Should I remove this line?

Comment (hageboeck) [reply roofit/roofit/src/RooDSCBShape.cxx]: Hmm, let me check again by building.

**Comment by hageboeck** [comment]
@phsft-bot build

File: roofit/roofit/src/RooDSCBShape.cxx (hunk 2 above)
Comment (hageboeck) [reply roofit/roofit/src/RooDSCBShape.cxx]: Oh, sorry. It's
```
#include <exception>
```

Comment (hageboeck) [reply roofit/roofit/src/RooDSCBShape.cxx]: I can push a commit to your branch, and we are good. Unless you want to get the praise, and have your name on the commit message. :-)

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by goi42** [comment]
@hageboeck Thanks for the correction. The build failure looks like it has something to do with a Git test, but it’s not clear to me what went wrong.

**Comment by hageboeck** [comment]
> @hageboeck Thanks for the correction. The build failure looks like it has something to do with a Git test, but it’s not clear to me what went wrong.

No worries, that's somebody else who messed up a setup. I think we are ready to merge, thanks for taking the time!

**Comment by hageboeck** [comment]
I'll wait for the CI to finish, and unless you can squash them into two commits that each add a single distribution, I will squash everything into a single commit. Github will only let you merge everything or squash everything, but that should be OK in this case. I don't expect that somebody would want to revert adding a distribution.

**Comment by goi42** [comment]
@hageboeck I will rebase into single commits for each distribution (two total). Thanks for your help!

**Comment by goi42** [comment]
@hageboeck Done. Let me know if you need any other changes!

**Comment by hageboeck** [comment]
@phsft-bot build

**Comment by SeungJunLee0** [comment]
you need to update or install the xcode.

**Pull request merged**